cartridge/twophase.lua

```diff
@@ -148,128 +187,140 @@ local function _clusterwide(patch)
             end
         end
+
+        if errmap ~= nil then
+            goto abort
+        else
+            goto apply
+        end
     end
 
-    if _2pc_error == nil then
-        -- 2PC commit
-        for _, uri in ipairs(configured_uri_list) do
-            local conn, err = pool.connect(uri)
-            if conn == nil then
-                log.error('Error commmitting config update at %s: %s', uri, err)
-                _2pc_error = err
-            else
-                local ok, err = errors.netbox_call(
-                    conn,
-                    '_G.__cluster_confapplier_commit_2pc'
-                )
-                if ok == true then
-                    log.warn('Committed config update at %s', uri)
-                else
-                    log.error('Error commmitting config update at %s: %s', uri, err)
-                    _2pc_error = err
-                end
+
+::apply::
+    do
+        log.warn('(2PC) Commit stage...')
+
+        local retmap, errmap = pool.map_call(
+            '_G.__cartridge_cwcfg_commit_2pc', nil,
+            {uri_list = uri_list, timeout = 5}
+        )
+
+        for _, uri in ipairs(uri_list) do
+            if retmap[uri] then
+                log.warn('Committed config at %s', uri)
             end
         end
-    else
-        -- 2PC abort
-        for _, uri in ipairs(configured_uri_list) do
-            if not configured_uri_list[uri] then
-                break
+        for _, uri in ipairs(uri_list) do
+            if retmap[uri] == nil then
+                local err = errmap and errmap[uri]
+                log.error('Error committing config at %s: %s', uri, err)
+                _2pc_error = err
             end
+        end
 
-            local conn, err = pool.connect(uri)
-            if conn == nil then
-                log.error('Error aborting config update at %s: %s', uri, err)
+        goto finish
+    end
+
+::abort::
+    do
+        log.warn('(2PC) Abort stage...')
+
+        local retmap, errmap = pool.map_call(
+            '_G.__cartridge_cwcfg_abort_2pc', nil,
+            {uri_list = abortion_list, timeout = 5}
+        )
+
+        for _, uri in ipairs(abortion_list) do
+            if retmap[uri] then
+                log.warn('Aborted config update at %s', uri)
             else
-                local ok, err = errors.netbox_call(
-                    conn,
-                    '_G.__cluster_confapplier_abort_2pc'
-                )
-                if ok == true then
-                    log.warn('Aborted config update at %s', uri)
-                else
-                    log.error('Error aborting config update at %s: %s', uri, err)
-                end
+                local err = errmap and errmap[uri]
+                log.error('Error aborting config update at %s: %s', uri, err)
             end
         end
+
+        goto finish
     end
 
+::finish::
     if _2pc_error == nil then
         log.warn('Clusterwide config updated successfully')
```

Comment:
Should it be logged after abort?

Reply:
No, and it won't be. Abort is a consequence of preparation error, which sets `_2pc_error ~= nil`

P.S. I fixed preparation phase accordingly to ensure that nasty box.NULL doesn't break the algorithm

Reply:
No of course, and it will not be. Preparation stage remembers _2pc_error, and it will do `log.error('Clusterwide config update failed')`